src/procedures/proc_bfs.c

```diff
@@ -0,0 +1,222 @@
+	if(SI_TYPE(args[0]) != T_NODE                 ||   // Source node.
+	   SI_TYPE(args[1]) != T_INT64                ||   // Max level to iterate to, unlimited if 0.
+	   !(SI_TYPE(args[2]) & (T_NULL | T_STRING))  ||   // Relationship type to traverse if not NULL.
+	   SI_TYPE(args[3]) != T_BOOL)                     // Whether traversed edges should be returned.
+		return PROCEDURE_ERR;
+
+	BFSContext *pdata = ctx->privateData;
+	pdata->depleted = false;
+	Node *source_node = args[0].ptrval;
+	GrB_Index source_id = ENTITY_GET_ID(source_node);
+	int64_t max_level = args[1].longval;
+	/* The BFS algorithm uses a level of 1 to indicate the source node. If this value is not
+	 * zero (unlimited), increment it by 1 to make level 1 indicate the source's direct neighbors. */
+	if(max_level > 0) max_level++;
+	const char *reltype = SIValue_IsNull(args[2]) ? NULL : args[2].stringval;
+	bool collect_edges = args[3].longval;
+
+	GraphContext *gc = QueryCtx_GetGraphCtx();
+	// Get edge matrix and transpose matrix, if available.
+	GrB_Matrix R;
+	GrB_Matrix TR;
+	if(reltype == NULL) {
+		R = Graph_GetAdjacencyMatrix(gc->g);
+		TR = Graph_GetTransposedAdjacencyMatrix(gc->g);
+	} else {
+		Schema *s = GraphContext_GetSchema(gc, reltype, SCHEMA_EDGE);
+		if(!s) return PROCEDURE_OK; // Failed to find schema, first step will return NULL.
+		pdata->reltype_id = s->id;
+		R = Graph_GetRelationMatrix(gc->g, s->id);
+		if(Config_MaintainTranspose()) TR = Graph_GetTransposedRelationMatrix(gc->g, s->id);
+		else TR = GrB_NULL;
+	}
+
+	GrB_Vector V;  // Vector of results
+	GrB_Vector PI = NULL; // Vector backtracking results to their parents.
+	GrB_Vector *PI_ptr = &PI;
+	if(!collect_edges) {
+		/* If we're not collecting edges, pass a NULL parent pointer so that the algorithm will
+		 * not perform unnecessary work. */
+		PI_ptr = GrB_NULL;
+		pdata->edges_output_idx = -1;
+	}
+	assert(LAGraph_bfs_both(&V, PI_ptr, R, TR, source_id, max_level, false) == GrB_SUCCESS);
+	/* Remove all values with a level less than or equal to 1.
+	 * Values of 0 are not connected to the source, and values of 1 are the source. */
+	GxB_Scalar thunk;
+	GxB_Scalar_new(&thunk, GrB_UINT64);
+	GxB_Scalar_setElement_UINT64(thunk, 1);
+	GxB_Vector_select(V, GrB_NULL, GrB_NULL, GxB_GT_THUNK, V, thunk, GrB_NULL);
+	GxB_Scalar_free(&thunk);
+
+	// Get number of entries.
+	GrB_Index nvals;
+	GrB_Vector_nvals(&nvals, V);
+	pdata->n = nvals;
+
+	// Retrieve all tuples representing connected nodes.
+	GrB_Index *node_ids = rm_malloc(nvals * sizeof(GrB_Index));
+	GrB_Vector_extractTuples_INT64(node_ids, GrB_NULL, &nvals, V);
+	pdata->node_ids = node_ids;
+	pdata->parents = PI;
+
+	GrB_Vector_free(&V);
+	return PROCEDURE_OK;
+}
+
+static SIValue *Proc_BFS_Step(ProcedureCtx *ctx) {
+	assert(ctx->privateData);
+
+	BFSContext *pdata = (BFSContext *)ctx->privateData;
+
+	// Return NULL if this source has already been mapped or there are no connected nodes.
```

Comment:
source has already been mapped is a bit unclear,
The BFS scan has already been performed at the invocation phase. I consider this as "mapping" the node.